### DISCUSSION TITLE
Editable DetailsList / Grid

### DISCUSSION THREAD

**Discussion by LukasSchauffele** [microsoft/powercat-creator-kit#162] [Q&A]
Hi All

I am keen to hear people's approaches when it comes to editing records in a list view (i.e., editable grids). I opted for a discussion post instead of a feature request as I couldn't find anything in the [Fluent UI Docs](https://developer.microsoft.com/en-us/fluentui#/controls/web/detailslist). However, there is a use case for being able to quickly edit multiple records. See the [Power Apps grid control](https://learn.microsoft.com/en-us/power-apps/maker/model-driven-apps/the-power-apps-grid-control) for Model Driven apps.

My quick workaround is to use the _DetailsList_ for the header components and use a standard gallery for the actual list. This is obviously not a great solution but it _sort of_ works for what I need it to, see below.

@scottdurow - are editable grids planned as part of the creator kit?

Thanks!

https://user-images.githubusercontent.com/7881214/197425771-8a972dc0-b324-4434-9d4f-c445826628e4.mp4

**Comment by yaseralhosani** [comment]
In DetailsList, an editable grid would be nice :)

Here is a [link](https://github.com/microsoft/powercat-creator-kit/issues) for requesting this feature.
New issue > Get started (Creator Kit Feature request)

**Comment by denise-msft** [comment]
Thanks for sharing your feedback @LukasSchauffele @yaseralhosani! (Very cool workaround)

It would be great if you could elaborate on your opinion for why the model driven app editable grid is not sufficient. Is the main concern that the grid should only take up part of the space in a form, rather than the whole page as a view? Would having an editable [subgrid](https://learn.microsoft.com/en-us/power-apps/maker/model-driven-apps/the-power-apps-grid-control#add-the-power-apps-grid-control-to-a-subgrid) achieve what you are looking for in that case? I'm making assumptions though, so please share more details if possible.

It's not currently on our backlog because of the product team-backed editable grid in model driven apps. That is the best implementation available right now and we would likely be duplicating a lot of work. If there are enhancements that are needed to make to the model driven app form meet your requirements so we can make this a configurable experience, that would be great feedback for us to find the best resolution.

**Reply by LukasSchauffele** [reply]
Hi @denisem-msft - thanks for your detailed reply. 

The Power Apps Grid Control is great for model driven apps. We are already using it, and are excited about getting more [features like grouping and aggregation](https://learn.microsoft.com/en-us/power-platform-release-plan/2022wave2/power-apps/grid-control-support-grouping-aggregation-nested-grids?WT.mc_id=BA-MVP-5004107). 

Unfortunately, this doesn't work for canvas apps. As far as I am aware it's not possible to embed a Power Apps Grid within a canvas app. 

I am looking for a solution _like the Power Apps Grid_ for canvas applications.

**Reply by denise-msft** [reply]
Hi @LukasSchauffele,

Glad to hear you are making good use out of the model driven app grid! It sounds like despite your satisfaction with the model driven grid itself, you still have scenarios that model driven apps cannot achieve, therefore you are asking for the editable grid in canvas here. The reason why I'm asking is because we are trying to architect Power Platform in a way that allows people to utilize the strengths of model and canvas (via pages) in complementary fashion, rather than duplicating the same functionality in both.

So, in this case, it would be very helpful if you can share more details around why you could not implement your scenario in a model driven form and why you'd prefer this be built in canvas (@yaseralhosani too). However, your post on this topic is a vote in it of itself and it's not the first we've heard, so only share more if you're comfortable. This additional information will just help us justify spending effort on this control if it's something we could not otherwise resolve in model driven apps.

**Reply by AstroNorman** [reply]
@denisem-msft  In my scenario, I use Canvas app instead of Model Driven App, because I use Sharepoint List as a data source.

And Model Driven app only support Dataverse tables. And Dataverse is going to cost additional 💸 in my department while Sharepoint List does not.

**Reply by LukasSchauffele** [reply]
@denisem-msft - For us it comes down to the _flexibility_ canvas apps provides. I'll use the app in my original post as an example. We are building an application to allow factory supervisors enter metrics about their site on a daily basis. Currently, they are using an Excel spreadsheet which works from an input perspective but creates double-up when transferring the data into a tabular format. They need to be able to enter this information quickly so it needs to be some a list rather than a form. Better yet, a table, with the individual dates as columns and the different groups and items as rows. Let's start with something simple and say we only care about 1 date (like what you can see in the video above). It should look like this:

|             |             | 5/12/2022 |
|-------------|-------------|-----------|
| Fabrication |             |           |
|             | Easy        | 5         |
|             | Med         | 5         |
|             | Complex     | 5         |
|             |             |           |
| Bays        |             |           |
|             | Rev Gen     | 5         |
|             | Non Rev Gen | 5         |
|             | Storage     | 5         |


In order to create this view in a model driven app, you would need to use filtering and grouping. Currently, this is only possible with the legacy editable grid and would look like this:

![CleanShot 2022-11-09 at 09 51 17](https://user-images.githubusercontent.com/7881214/200693227-c46f8117-0736-4284-bd6f-bf802868ba79.png)


This view is _okay_; some of the issues:

- Can't easily identify which dates I am looking at (unless I show it on each record)
- Can't easily switch between dates. Need to use the filtering option
- Can't easily compare between two dates
- Group by doesn't save so it needs to be set every time

Now, let's say i have a further requirement to show multiple days along the top, as my columns, so it would look like this:

|             |             | 5/12/2022 | 12/12/2022 | 19/12/2022 | 26/12/2022 | 3/1/2023 |
|-------------|-------------|-----------|------------|------------|------------|----------|
| Fabrication |             |           |            |            |            |          |
|             | Easy        | 5         | 4          | 3          | 2          |          |
|             | Med         | 5         | 4          | 3          | 2          |          |
|             | Complex     | 5         | 4          | 3          | 2          |          |
|             |             |           |            |            |            |          |
| Bays        |             |           |            |            |            |          |
|             | Rev Gen     | 5         | 4          | 3          | 2          |          |
|             | Non Rev Gen | 5         | 4          | 3          | 2          |          |
|             | Storage     | 5         | 4          | 3          | 2          |          |


Not sure if this could be displayed in a model driven app. However, I can solve these issues in a canvas app because I have complete control over my table. 

Maybe there are some other controls or plugins I don't know of which would let me achieve this in a model driven app. If so, I am all ears 😄 

Mentioning @scottdurow as I know he's been working with editable grids.

**Reply by LenCardinal** [reply]
@denisem-msft, your comment "utilize the strengths of model and canvas (via pages) in complementary fashion" was intriguing.  I also noticed that the MDA sample app in the creator's kit was all custom (canvas) pages - the hybrid approach I think you were referring to.  I'm trying to determine direction for my company as we set out adopting PowerApps.  I get the sense that MDA with Custom pages is the preferred direction (is that the case?).

But I'm only noticing this approach out of my peripheral vision.  The real question is: how do I keep up with Microsoft's guidance and direction?  I haven't found any blogs, forums, etc that discuss strategy and futures. Can you point me to any?  Forums, blogs, conference archives, etc?

Thanks.
